### ISSUE TITLE
Project Plan: Your Pocket Therapist

### ISSUE THREAD

**Issue by vivianknee** [vivianknee/PocketTherapist#1]
# Project Plan: Your Pocket Therapist
Your Pocket Therapist uses an AI Face algorithm to accurately understand the user's emotions and provide advice/quotes in response.

- Nighthawk pages frontend
- Spring portfolio backend

## Team Members
<br>
Vivian N
Isabelle G
Tanisha P
Sreeja G

## Project Objectives
1. Create a user-friendly interface
2. Create a live emotion detector, if make it so photos can be uploaded
3. Utilize a Face API to detect images and output emotions
4. Create our own unique database of advice and quotes that correspond to the emotions
5. Focus on JS for the frontend and Java for the backend

## Project Tasks and Timeline
1. Frontend (JS HTML)
- [x] HTML Formatting (overall sass)
- [x]  Container for Live Camera
- [x]  Section for Emotion recognizer result(with corresponding emoji), percent accuracy
- [x]  Section for a generated quote (From API/Database call)

2. Emotions AI (PYTHON)
- [x] Open CV integration to live data collection
- [x]  Find API/Data for the model -- RapidAPI
- [x]  Training/Testing
- [x]  Display emotion result, save to global variable
- [x]  Percent accuracy calculation

3. Quote Generation (JAVA)
- [x]  Access quote based on saved emotion result
- [ ]  FIND API
- https://rapidapi.com/karanp41-eRiF1pYLK1P/api/world-of-quotes
- [x]  Display successful api/database call

**Comment by tanishapatil1234** [comment]
## Sub ticket Tanisha CNN implementation 

###Week 1: Planning and Data Preparation

**Day 0: Sprint Planning**

- [ ] - Define the project scope and objectives.
- [ ] - Create user stories and product backlog.
- [ ] - Set up a Scrum board or Jira project.
- [ ] - Assign roles and responsibilities: Product Owner, Scrum Master, Development Team.
- [ ] - Kick-off meeting to clarify goals and expectations.

**Day 1: Data Collection**

- [ ] Gather a dataset of facial images labeled with emotions (e.g., Kaggle's FER-2013 dataset).
- [ ] Clean the dataset by removing irrelevant or low-quality images.

**Day 2: Data Preprocessing**

- [ ] Resize images to a consistent format (e.g., 224x224 pixels).
- [ ] Normalize pixel values to a common scale (e.g., [0, 1]).
- [ ] Split the dataset into training, validation, and test sets.
- [ ] Augment the training dataset (e.g., rotate, flip, zoom) to increase diversity.

**Day 3: Model Architecture Design**

- [ ] Decide on a CNN architecture (e.g., VGG16, ResNet, or custom architecture).
- [ ] Define the layers, filters, and activation functions for the model.
- [ ] Document the model architecture on your Scrum board.

**Day 4: Model Implementation**

- [ ] Start coding the CNN model in a deep learning framework like TensorFlow or PyTorch.
- [ ] Implement forward and backward passes, loss function, and optimization algorithm (e.g., Adam).
- [ ] Ensure the model can handle the input image size and number of classes.

**Day 5: Training**

- [ ] Train the model on the preprocessed dataset.
- [ ] Monitor training metrics such as loss and accuracy.
- [ ] Use early stopping if validation performance stalls.
- [ ] Document training progress on your Scrum board.

**Day 6: Model Evaluation**

- [ ] Evaluate the model's performance using the test set.
- [ ] Calculate metrics like accuracy, precision, recall, and F1-score.
- [ ] Identify potential sources of bias or error and document them.

###Week 2: Model Optimization and Deployment

**Day 7: Hyperparameter Tuning**

- [ ] Fine-tune hyperparameters, like learning rate, batch size, or dropout rates.
- [ ] Experiment with different CNN architectures to improve performance.

**Day 8: Error Analysis**

- [ ] Analyze model errors on the test set to understand common misclassifications.
- [ ] Adjust the model or dataset to address error patterns.

**Day 9: Model Interpretability**

- [ ] Implement visualization techniques (e.g., Grad-CAM) to understand what the model focuses on when making predictions.

**Day 10: Model Optimization**

- [ ] Implement model optimization techniques (e.g., weight quantization or model pruning) for deployment on resource-constrained devices.

**Day 11: Deployment Setup**

- [ ] Set up the infrastructure and environment for model deployment (e.g., cloud server or edge device).
- [ ] Create an API or interface for the model.

**Day 12: Testing and Validation**

- [ ] Test the deployed model to ensure it functions correctly in the production environment.
- [ ] Validate the model's real-world performance.

**Day 13: Documentation and Training**
- [ ] Document the model architecture, training process, and deployment steps.
- [ ] Train end-users or colleagues on using the model effectively.

**Day 14: Final Review and Retrospective**

- [ ] Review the project's achievements and ensure all requirements are met.
- [ ] Conduct a retrospective meeting to discuss what went well and what could be improved in the development process.
- [ ] Plan for future iterations or improvements.

**Comment by vivianknee** [comment]
## Plan

- on quotes.md, use a post-JS to post new quotes to the database. Ensure that when the page is refreshed, the added quote still remains. 
- gallery.md will also have JS code to format and display each quote with HTML and CSS in a more efficient way
<img width="456" alt="Screenshot 2023-10-25 at 12 14 04 PM" src="https://github.com/vivianknee/PocketTherapist/assets/82780605/7f26e298-e642-426b-ad8b-34abb4460c53">


 ### issues
- ran into a cors issue. had to add this to nginx file. Now I can locally test the post
<img width="611" alt="Screenshot 2023-10-25 at 12 15 08 PM" src="https://github.com/vivianknee/PocketTherapist/assets/82780605/d8d207fb-2b79-4ec0-bcb9-177b3ec5200d">

- on gallery.md, use a get fetch call to update the page with all the new quotes, including quotes that were added on quotes.md. 
<img width="457" alt="Screenshot 2023-10-25 at 12 15 23 PM" src="https://github.com/vivianknee/PocketTherapist/assets/82780605/12dc3403-26c5-4f99-9998-a192ea3706b7">

**Comment by sreejagangapuram** [comment]
**SUB REVIEW TICKET: CAMERA FEATURE (SREEJA)**

- Camera progress  
<img width="1440" alt="Screen Shot 2023-10-25 at 12 13 08 PM" src="https://github.com/vivianknee/PocketTherapist/assets/111466890/b82b3499-aab3-4773-b55d-edae0ccb988e">

- Code for Camera
MD file 
<img width="898" alt="Screen Shot 2023-10-25 at 12 14 24 PM" src="https://github.com/vivianknee/PocketTherapist/assets/111466890/1a04eef9-f5ad-4440-92b6-fb3dba37e591">
<img width="930" alt="Screen Shot 2023-10-25 at 12 14 31 PM" src="https://github.com/vivianknee/PocketTherapist/assets/111466890/1fcb982e-c2ba-4de9-a1a8-9594e99a7565">

Java Script Code 
<img width="666" alt="Screen Shot 2023-10-25 at 12 15 22 PM" src="https://github.com/vivianknee/PocketTherapist/assets/111466890/15664ddc-415e-414e-85b5-350adf70ae2f">

- Commits

**Comment by isabelle926** [comment]
### AI Chatbot (Isabelle)
**Plan: **
create chatbot feature that uses conversations from real therapists to emulate a therapist for the user

Frontend:
<img width="1317" alt="Screen Shot 2023-10-25 at 12 19 19 PM" src="https://github.com/vivianknee/PocketTherapist/assets/70926137/b45b58bb-578a-4f4c-a6bb-3b027cc70e19">

Backend: 
* use Apache OpenNLP

![image](https://github.com/vivianknee/PocketTherapist/assets/70926137/badddda3-75f4-4832-ac7a-6b30bd5d3bc0)

Issues:
* Use AIML for backend instead 
* Connect frontend to backend
